themes/default/content/blog/iac-recommended-practices-developer-stacks-git-branches/index.md

```diff
@@ -0,0 +1,209 @@
+
+### Using developer stacks
+
+We've seen a few ways that having multiple developers share a single stack for ad-hoc deployment can lead to various problems. But we've also seen at least one situation in which being able to run ad-hoc deployments --- without the risk of colliding with existing environments --- can be quite valuable. In fact there are many reasons you might want to do this --- and fortunately Pulumi makes it quite easy to do so.
+
+The recommended practice is to allow developers to create and manage their own Pulumi stacks. In this workflow, developers working on feature branches are able to iterate freely on both application and infrastructure, deploying as often as necessary into a shared cloud environment (typically a sandbox environment, isolated from production) and destroying cloud infrastructure when it's no longer needed.
+
+For the developer, the typical workflow looks something like this:
+
+1. Check out a new feature branch
+2. Make changes to application and/or infrastructure code
+3. Select your personal developer stack with `pulumi stack select`
+4. Deploy your stack into your team's shared cloud environment with `pulumi up`
+5. Commit your code and submit a pull request, optionally including a link either to the Pulumi update (e.g., when using the Pulumi Service) or to certain key resources belonging to the stack --- for example, an application's primary HTTP endpoint
+
+The benefits of a workflow like this one are many:
+
+* No more resource-naming collisions. Because Pulumi names cloud resources [uniquely by default](/docs/intro/concepts/resources/names/#autonaming), no two stacks will ever produce cloud resources with the same names.
+* No more deployment conflicts. When every developer has a stack of their own, they can deploy to the sandboxed environment as often as necessary, without having to wait for another deployment to finish first.
+* Everyone on the team has their own production-like environment to use for end-to-end testing.
+
+For Zephyr, the quickest path to adopting this workflow was to use the `dev` stack as a foundation for new developer stacks. To do that, team members simply ran the following Pulumi commands:
+
+```bash
+pulumi stack init zephyr/developer-name --copy-config-from zephyr/dev
+pulumi up
+```
+
+Developer stacks are considered ephemeral, so they're torn down with `pulumi destroy` when not in use in order to cut down on cost. Stack configuration files (e.g., `Pulumi.developer-name.yaml`) are checked into version control and stored alongside other stacks like `Pulumi.dev.yaml` and `Pulumi.prod.yaml`.
+
+### Keeping feature-branch workflows
+
+We've also seen that the Zephyr team's existing Git workflow (short-lived feature branches merged frequently into a long-lived base branch) has worked well for them --- and it turns out this workflow also works well when combined with the use of developer stacks.
```

Comment:
what i am not quite understanding is do separate branches in git automatically spin up an separate stack? 

like is this a manual process they are folllowing or is this somehow built into automation?

Reply:
>do separate branches in git automatically spin up an separate stack

They don't; there's no technical association between current Git branch and current Pulumi stack. You have to `pulumi stack select` manually.

Do you think we should do more to explain this? Or maybe the docs could?

Reply:
There's a lot of depth here, in my opinion; it wouldn't hurt to explore this in more depth in a separate (non-Zephyr) blog post, I think. Some graphics perhaps, to help explain it. (Although, TBH, I'm not sure how to graphically represent this.)

Reply:
Chatted with the team and we're going to add a graphic for this as a follow-up.